### ISSUE TITLE
Newlines included with `ignore_hidden_elements = true` but not without?

### ISSUE THREAD

**Issue by henrik** [teamcapybara/capybara#2282]
This confused me – not sure if it counts as a bug, but it seems unintuitive if nothing else.

Capybara 3.29.0, running a feature spec with RackTest on a page that does `<p>foo</p> <p>bar</p>`.

With `ignore_hidden_elements = false`,  a spec like `expect(page).to have_content "foo bar"` passes.

But with `ignore_hidden_elements = true`,  it fails because the content is now seen as "foo\nbar".

Another similarly unintuitive difference is that something like `<span style="text-transform: uppercase">hello</span>` would pass `have_content("hello")` before, but now we need `"HELLO"`.

Maybe I missed it in the docs, but if not, maybe these non-obvious consequences of the setting could be explained there?

**Comment by henrik** [comment]
This (at least the newline bit) seems related to the `normalize_ws` setting. I guess (but haven't confirmed) that with `ignore_hidden_elements = true`, you automatically get the old default of `normalize_ws` being true.

**Comment by twalpole** [comment]
TL;DR; Don't set `ignore_hidden elements` to false when testing

When you set `ignore_hidden_elements` to false (which you should never do when testing, has potentially valid use cases when doing web scraping though)- It changes the default behavior of `text` to return all text (not just visible text).  When the rack_test driver is determining visible text to return it looks at the types of child elements inside the element you're getting the text from and will replace block elements with `\n`s to replicate what happens in a real browser. However when returning all text (not just visible) it doesn't do that because it has no way to know how any of it would be rendered.  This is as designed.  

As for your `text-transform`  issue - I would expect that in a driver that uses a real browser since the intention is for that to match the text as displayed as closely as possible - I would not expect that when using the rack_test driver because it's not a real browser, doesn't actually render the page, and only attempts to mimic a VERY limited subset of in element defined CSS styles  (generally around visibility)

**Comment by henrik** [comment]
Thank you, @twalpole! Really appreciate you taking the time. 

We’re moving a project from the “false” to the “true” setting, so agreed on not wanting to use the former.

That explanation for the newlines makes perfect sense. It was still a bit of a gotcha - changing the visibility bool also causing specs to fail from newlines being introduced. But I know we’re way behind the curve with this transition, so maybe spending time documenting that gotcha better would be low value. I might still see if I can think of a PR that would have helped clarify it for me, though.

And you’re right about the caps thing not being RackTest. I edited that example in later when I ran across it in an Apparition test and forgot I explicitly mentioned RackTest earlier. But I did see test failures where I had to capitalise some expected text after toggling the ignore_hidden_elements config to true. Any theories on why that config would have this result? I’ll double-check this tomorrow at work and make sure I can reproduce it by just toggling this config.

**Comment by twalpole** [comment]
@henrik Yes - it's basically the same reason - When looking for only visible text (which is the default when ignore_hidden_elements is true) Apparition calls `HTMLElement.innerText` which returns the text as rendered - when set to also return non-visible text (the default when ignore_hidden_element is false) it has to use `HTMLElement.textContent` which is not aware of rendering.

**Comment by henrik** [comment]
Ah! Thank you so much again.
